localstack/services/kms/kms_listener.py

```diff
@@ -1,25 +1,219 @@
+        del grants[data[GRANT_ID]]
+    elif "GrantToken" in data:
+        KMSBackend.get().grants = {
+            grant_id: grant
+            for grant_id, grant in grants.items()
+            if data["GrantToken"] not in grant[GRANT_TOKENS]
+        }
+    else:
+        raise ValidationException("Grant token OR (grant ID, key ID) must be specified")
+    return {}
+
+
+def handle_revoke_grant(data):
+    grants = KMSBackend.get().grants
+
+    if GRANT_ID in data and KEY_ID in data and grants[data[GRANT_ID]][KEY_ID] == data[KEY_ID]:
+        del grants[data[GRANT_ID]]
+        return {}
+    else:
+        raise ValidationException("Grant ID, key ID must be specified")
+
+
+def handle_list_retirable_grants(data):
+    if RETIRING_PRINCIPAL not in data:
+        raise ValidationException("Retiring principal must be specified")
+
+    grants = KMSBackend.get().grants
+
+    limit = data.get("Limit", 50)
+
+    if "Marker" in data:
+        markers = KMSBackend.get().markers
+        filtered = markers.get(data["Marker"], [])
+    else:
+        filtered = [
+            grant
+            for grant in grants.values()
+            if RETIRING_PRINCIPAL in grant and grant[RETIRING_PRINCIPAL] == data[RETIRING_PRINCIPAL]
+        ]
+    if len(filtered) <= limit:
+        return {"Grants": filtered, "Truncated": False}
+
+    markers = KMSBackend.get().markers
+
+    in_limit = filtered[:limit]
+    out_limit = filtered[limit:]
+
+    marker_id = long_uid()
+    markers[marker_id] = out_limit
+
+    return {"Grants": in_limit, "Truncated": True, "NextMarker": marker_id}
+
 
 class ProxyListenerKMS(ProxyListener):
     def forward_request(self, method, path, data, headers):
         action = headers.get("X-Amz-Target") or ""
-        if action.endswith(".CreateKey"):
-            data1 = json.loads(to_str(data))
-            descr = data1.get("Description") or ""
-            event_publisher.fire_event(EVENT_KMS_CREATE_KEY, {"k": event_publisher.get_hash(descr)})
+        if method == "POST" and path == "/":
+            parsed_data = json.loads(to_str(data))
+
+            if action.endswith(".CreateKey"):
+                descr = parsed_data.get("Description") or ""
+                event_publisher.fire_event(
+                    EVENT_KMS_CREATE_KEY, {"k": event_publisher.get_hash(descr)}
+                )
+            elif action.endswith(".CreateGrant"):
+                return handle_create_grant(parsed_data)
+            elif action.endswith(".ListGrants"):
+                return handle_list_grants(parsed_data)
+            elif action.endswith(".RevokeGrant"):
+                return handle_revoke_grant(parsed_data)
+            elif action.endswith(".RetireGrant"):
+                return handle_retire_grant(parsed_data)
+            elif action.endswith(".ListRetirableGrants"):
+                return handle_list_retirable_grants(parsed_data)
         return True
 
 
+class KMSBackend(RegionBackend):
+    def __init__(self):
```

Comment:
nit: We could add type hints and a short description to these class fields, something like:
```
class KMSBackend(RegionBackend):
    # maps grant IDs to grant details
    grants: Dict[str, Dict]
    # maps marker strings to paginated result lists
    markers: Dict[str, List]

    def __init__(self):
        self.grants = {}
        self.markers = {}
```